packages/parbox.lua

```diff
@@ -0,0 +1,498 @@
+      vboxlist[i].depth = d -- Hack depth of first vbox
+      break
+    end
+  end
+end
+
+local parseBorderOrPadding = function (rawspec, opt)
+  local spec
+  if type(rawspec) == "table" then
+    spec = rawspec
+  else
+    spec = {}
+    for token in SU.gtoke(rawspec, "[ ]+") do
+      if(token.string) then
+        local value = SU.cast("measurement", token.string)
+        spec[#spec+1] = value:tonumber()
+      end
+    end
+  end
+  if #spec == 1 then
+    return { spec[1], spec[1], spec[1], spec[1] }
+  end
+  if #spec ~= 4 then SU.error("Invalid "..opt.." specification: "..rawspec) end
+  return spec
+end
+
+SILE.registerCommand("parbox", function (options, content)
+  local width = SU.required(options, "width", "parbox")
+  local strut = options.strut or "none"
+  local border = options.border and parseBorderOrPadding(options.border, "border") or { 0, 0, 0, 0 }
+  local valign = options.valign or "top"
+  local padding = options.padding and parseBorderOrPadding(options.padding, "padding") or { 0, 0, 0, 0 }
+  local bordercolor =  options.bordercolor and SILE.colorparser(options.bordercolor)
+  local minimize = SU.boolean(options.minimize, false)
+
+  width = SILE.length(SU.cast("measurement", width)):absolute()
+
+  local vboxes = parboxFraming({ width = width }, content)
+
+  local strutDimen
+  if strut == "rule" then
+    strutDimen = SILE.call("strut", { method = "rule" })
+    insertStruts(vboxes, strutDimen)
+  elseif strut == "character" then
+    strutDimen = SILE.call("strut", { method = "character" })
+    insertStruts(vboxes, strutDimen)
+  else
+    strutDimen = { height = SILE.length(0), depth = SILE.length(0) }
+  end
+
+  local wmax = SILE.length()
+  local totalHeight = SILE.length()
+  for i = 1, #vboxes do
+    -- Try to cancel vertical stretching/shrinking
+    if vboxes[i].is_vglue then
+      -- Important: many vglues are just the _same_ node, which will be "adjusted"
+      -- by the page builder. We cannot tweak directly its height or depth as we
+      -- sometimes do with other boxes, as it would have a side effect. So we have
+      -- to re-create a new vglue with the appropriate fixed dimension.
+      vboxes[i] = SILE.nodefactory.vglue(SILE.length(vboxes[i].height.length))
+    end
+    totalHeight = totalHeight + vboxes[i].height:absolute() + vboxes[i].depth:absolute()
+
+    if minimize then
+      -- We go through all nodes in the line to compute the maximal line width.
+      -- This is not so efficient, we could have modified the typesetter to
+      -- compute it when it was processing the lines...But is is buried deep
+      -- in its breakpointsToLines() method, which some packages may moreover
+      -- overload, so we went for the easiest solution, albeit CPU-intensive.
+      local w = SILE.length()
+        if vboxes[i].nodes then
+          for n in ipairs(vboxes[i].nodes) do
+            if vboxes[i].nodes[n].width > SILE.length(0) then
```

Comment:
This is too naive and doesn't minimize the parbox as much as possible: negative width can of course occur (e.g. via kerns, or lskip/rskip/indent) and shouldn't be skipped.

I was all too quick here. I just noticed that recently... FWIW, the fun part of the story is that I liked so much the book example shown by @alerque in #911 that I decided to use it for a "fancy" custom table of contents with 2 levels (parts and chapters)... and met that case. I'll propose a fix and for a better width computation. The fix sounds decent now, with the nice TOC I had in mind, but I'll of course want to check it more before submitting it, and I'll also extract a dedicated test:
![image](https://user-images.githubusercontent.com/18075640/162632694-15d17986-c3f5-4e8f-9994-0b3639ba873b.png)